src/oracle/middleware.js

```diff
@@ -0,0 +1,761 @@
+    leader,
+    {
+      // @ts-expect-error xxx
+      unserializer,
+    },
+  );
+
+  //get offers
+  let offers = await getOffers(follower)
+
+  return Number(offers[0]["status"]["invitationSpec"]["invitationArgs"][0]["roundId"])
+}
+
+/**
+ * Function to check if submission was satisfied for a specific round
+ * @param {*} oracle oracle address
+ * @param {*} feedOfferId the offer id of the feed to check for
+ * @param {*} roundId the round Id which is checked
+ * @returns whether a submission was successful for a specific round
+ */
+export const checkSubmissionForRound = async (oracle, feedOfferId, roundId) => {
+  const unserializer = boardSlottingMarshaller(fromBoard.convertSlotToVal);
+  const leader = makeLeader(networkConfig.rpcAddrs[0]);
+
+  const follower = await makeFollower(
+    `:published.wallet.${oracle}`,
+    leader,
+    {
+      // @ts-expect-error xxx
+      unserializer,
+    },
+  );
+
+  //get offers
+  let offers = await getOffers(follower)
+  
+  //loop through offers starting from last offer
+  for (var i = 0; i < offers.length; i++) {
+    //get current offer
+    var currentOffer = offers[i];
+
+    //if a price invitation and for the correct feed
+    if (currentOffer["status"]["invitationSpec"]["invitationMakerName"] == "PushPrice" && currentOffer["status"]["invitationSpec"]["previousOffer"] == feedOfferId) {
+      let offerRound = Number(currentOffer["status"]["invitationSpec"]["invitationArgs"][0]["roundId"])
+
+      //if it is an offer for the round we are checking for
+      if (offerRound == roundId) {
+        //if there is no error
+        if (!currentOffer["status"].hasOwnProperty("error")) {
+          return true
+        }
+      }
+      //else if offer round id is less than the round we want to check and its satisfied
+      //return false because there cannot be a submission for a newer round before this offer 
+      else if (offerRound < roundId && !currentOffer["status"].hasOwnProperty("error")) {
+        return false
+      }
+    }
+  }
+  return false
+}
+
+/**
+  * Function to query price from chain
+  * @param {*} feed feed name of the price to query in the form of ATOM-USD
+  * @returns the latest price
+  */
+export const queryPrice = async (feed) => {
+
+  try {
+    //read value from vstorage
+    const capDataStr = await readVStorage(vstorage, feed, false);
+
+    //parse the value
+    var capData = JSON.parse(JSON.parse(capDataStr).value)
+    capData = JSON.parse(capData.values[0])
+    //replace any extra characters
+    capData = JSON.parse(capData.body.replaceAll("\\", ""))
+
+    //get the latest price by dividing amountOut by amountIn
+    var latestPrice = Number(capData.amountOut.value.digits) / Number(capData.amountIn.value.digits)
+
+    console.log(feed + " Price Query: " + String(latestPrice))
+    return latestPrice
+  }
+  catch {
+    return 0
+  }
+}
+
+/**
+ * Function to get oracles feed invitations
+ * @returns an object containing feed invitation IDs
```

Comment:
Can you specify what the object is containing i.e. the struct

Reply:
Cannot specify exactly since the fields will be the feed names but explained this briefly in https://github.com/jacquesvcritien/agoric-cl-middleware/commit/822e96128b23515cb3b7a9232e567e8c671dfe15